### PR TITLE
Slack: interrupt streaming after timeout on user action

### PR DESCRIPTION
**Pull request by spolu**
## Description

Fixes: https://github.com/dust-tt/tasks/issues/7990
Fixes: https://github.com/dust-tt/tasks/issues/7848
Context: https://app.dust.tt/w/0ec9852c2f/conversation/1F42d4Jbo3

Before this change when a conversation was blocked on a user action that was not taken, the streaming would timeout at 5mn and retry 10 times, and finally error after 50mn with message:

`An unexpected error occurred while answering your message, please retry.`

First:

Screenshot 2026-05-12 at 14 14 24

And then:

Screenshot 2026-05-12 at 14 13 32

We introduce a timeout handler plugged into the pre-existing cancellation mechanism of the streaming SDK to have a smooth cancellation.

New behaviour while not perfect (action cannot be taken from Slack after timeout) we have an informative error message and an invitation to continue the conversation on Dust.

First:

Screenshot 2026-05-12 at 14 42 10

And then:

Screenshot 2026-05-12 at 14 45 05

And if the user clicks on one of the buttons:

Screenshot 2026-05-12 at 14 45 32

Limitation: we cannot remove the approval call to action post timeout as ephemeral elements cannot be removed without knowing their `response_url`, which is only received once a user take action. If the user on which the approval is pending clicks on it (others don't see it) the small ⚠️ icon is shown to report an error taking action.

This should remove all these errors [Errors triggering retries](https://app.datadoghq.eu/logs?query=%22Failed%20processing%20event%20stream%22%20service%3Aconnectors&agg_m=count&agg_m_source=base&agg_t=count&cols=host%2Cservice&event=AwAAAZ4XgJt5rFX8GQAAABhBWjRYZ0txbEFBRGdLUjFVay1PTWh3QWMAAAAkMDE5ZTE3OGItOThiZC00ODFkLTk0OTktMWQ2OTk3MzA4ZjU2AAAY7A&messageDisplay=inline&refresh_mode=sliding&storage=hot&stream_sort=desc&viz=stream&from_ts=1775920467523&to_ts=1778512467523&live=true) and [Errors post retries exhausted](https://app.datadoghq.eu/logs?query=%22Unexpected%20exception%20answering%20to%20Slack%20Chat%20Bot%20message%22%20%22Error%3A%20Exceeded%20maximum%20reconnection%20attempts%22&agg_m=count&agg_m_source=base&agg_t=count&clustering_pattern_field_path=message&cols=host%2Cservice%2C%40connectorId&link_event_id=8618925699268577964&link_event_ts=1777990810&link_monitor_id=22042330&link_source=monitor_notif&messageDisplay=inline&refresh_mode=sliding&storage=hot&stream_sort=desc&viz=stream&from_ts=1775919532277&to_ts=1778511532277&live=true). Example [Trace](https://app.datadoghq.eu/apm/trace/69fc6fec0000000029bc73cf0ea79419?spanID=2217120249777622841&timeHint=1778151404692).

## Tests

Tested locally

## Risk

Low corner case handling.

## Deploy Plan

- deploy `connectors`

### CODE REVIEW

**Comment by ghost** [comment]
Ready to review this PR? Stage has broken it down into 4 individual chapters for you:

| | Title |
|---|-------|
| 1 | [Define Slack user action timeout types and utilities](https://stagereview.app/dust-tt/dust/pull/25508/chapters/1) |
| 2 | [Implement timeout management logic and state](https://stagereview.app/dust-tt/dust/pull/25508/chapters/2) |
| 3 | [Integrate timeout triggers into event stream](https://stagereview.app/dust-tt/dust/pull/25508/chapters/3) |
| 4 | [Handle timeout completion and Slack notification](https://stagereview.app/dust-tt/dust/pull/25508/chapters/4) |

<a href="https://stagereview.app/dust-tt/dust/pull/25508">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://stagereview.app/assets/gh-open-in-stage-dark.svg">
    Open in Stage
  </picture>
</a>

<sub>Chapters generated by <a href="https://stagereview.app">Stage</a> for commit aff9ab20d5cf11be841aba6476aaa7696d2701d2 on May 12, 2026 1:41pm UTC.</sub>

File: connectors/src/connectors/slack/chat/stream_conversation_handler.ts [hunk 1]
```diff
@@ -191,6 +192,45 @@ class SlackAnswerRetryableError extends Error {
...
+      actionLabel = "file access authorization";
+      break;
+    case "tool_personal_auth_required":
+      actionLabel = "tool authentication";
+      break;
+    case "tool_ask_user_question":
+      actionLabel = "a response to a question";
+      break;
+    default:
+      assertNever(actionType);
```
[the remark this answers is not in the document]
Comment (spolu) [reply connectors/src/connectors/slack/chat/stream_conversation_handler.ts:224]: This is not client code.

**Comment by spolu** [comment]
Improved the wording a bit to clarify that things have to happen on Dust:

Screenshot 2026-05-12 at 14 56 23

File: connectors/src/connectors/slack/chat/stream_conversation_handler.ts [hunk 2]
```diff
@@ -242,10 +290,73 @@ async function streamAgentAnswerToSlack(
     },
   };
 
-  for await (const event of streamRes.value.eventStream) {
+  function clearUserActionTimeout() {
```
Comment (flvndvd) [inline connectors/src/connectors/slack/chat/stream_conversation_handler.ts]: `nit` Let's avoid nesting functions, better:


```suggestion
  const clearUserActionTimeout = () => {
```

File: connectors/src/connectors/slack/chat/stream_conversation_handler.ts [hunk 3]
```diff
@@ -242,10 +290,73 @@ async function streamAgentAnswerToSlack(
...
+        conversationId: conversation.sId,
+        pendingActionType: pendingUserActionType,
+      },
+      "Clearing user-action idle timeout for Slack stream."
+    );
+
+    pendingUserActionType = null;
+  }
+
+  function startUserActionTimeout(actionType: SlackUserActionType) {
```
Comment (flvndvd) [inline connectors/src/connectors/slack/chat/stream_conversation_handler.ts]: Same here.

**Review by flvndvd — APPROVED** [review]

File: connectors/src/connectors/slack/chat/stream_conversation_handler.ts (hunk 2 above)
Comment (spolu) [reply connectors/src/connectors/slack/chat/stream_conversation_handler.ts]: 👍

File: connectors/src/connectors/slack/chat/stream_conversation_handler.ts (hunk 3 above)
Comment (spolu) [reply connectors/src/connectors/slack/chat/stream_conversation_handler.ts]: 👍
